[PyTorch] Change representation of SizesAndStrides

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47508 [PyTorch] Change representation of SizesAndStrides**
* #47507 [PyTorch] Introduce packed SizesAndStrides abstraction

This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

swolchok committed:
[PyTorch] Change representation of SizesAndStrides

This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
add more test coverage and fix resize bug on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
rebase on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

ezyang commented:
I know the point of this change isn't really to affect instruction count, but I'm curious to know which way instruction count on the microbenchmarks goes this way. I'm sort of expecting count to go up but maybe I will be surprised.

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -73,46 +194,120 @@ class C10_API SizesAndStrides {
+  }
+
+  static constexpr int MAX_INLINE_SIZE = 5;
+  int64_t taggedStorageOrSize_;
```
nit: wouldn't `intptr_t` be better here?

ezyang replied:
NB: when it's a pointer, this is `std::vector<int64_t>* taggedStorageOrSize_`, which means there's actually a *double indirection* when you overflow the inline fields. I can see using a `std::vector` makes implementation a lot more convenient, so maybe it's OK for now, but while the patient is open it might be best to do "the best we can" implementation-wise.

swolchok replied:
> while the patient is open it might be best to do "the best we can" 

This runs counter to the philosophy that diffs should be small and focused. If performance in the out-of-line case is important for some reason, we can spend a second diff focusing on that. (But if it's important, that invalidates the assumption underlying this implementation -- we are intentionally optimizing rank <= 5 at the expense of larger ranks!)

swolchok replied:
That said, I'll switch representations as I mentioned in the other subthread because it becomes easy to get rid of the vector.

ezyang replied:
> This runs counter to the philosophy that diffs should be small and focused. If performance in the out-of-line case is important for some reason, we can spend a second diff focusing on that. 

Acknowledged. I'm mostly looking for opportunities along the lines of "it's the same net amount of work, but is ever so slightly more optimized." If it's a lot more work definitely OK for you to  push back like that (though, I'm glad you liked my other suggestion!)

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -73,46 +194,120 @@ class C10_API SizesAndStrides {
+  }
+
+  void setInlineSize(size_t sz) {
+    taggedStorageOrSize_ = (sz << 1) | 1;
```
note to self: leaky leaky

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -17,26 +30,118 @@ class C10_API SizesAndStrides {
+    }
+    if (C10_LIKELY(rhs.isInline())) {
+      if (C10_UNLIKELY(!isInline())) {
+        delete outOfLineStorage();
```
This is all hella manual but I guess we don't have a choice if we're going to go the tagged pointer route.

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -8,6 +10,17 @@ namespace c10 {
+// is 0, the tag word is a pointer to an array of int64_t holding the
+// data; the first int64_t indicates the size of the rest.
+// 5 int64_t reserved for inline sizes
+// 5 int64_t reserved for inline strides
```
I liked the tagged bit proposal when you first proposed it, but after taking a closer look at this PR, I was thinking: why do we need a tagged pointer at all? Make the first word the size unconditionally; when size is greater than 5, stash the pointer in the inline fields. When we store out of line the inline space is totally wasted, so might as well just use it to store the pointer. (You could even inplace construct the vector here, so you don't have to reimplement an out-of-line constructor).

The benefit of the alternate encoding is that you can now retrieve the dim() of a tensor without any branches. The net number of branches stays the same. I guess one downside is some of the bookkeeping is more complicated since you have to avoid clobbering data you might need for later in the inline fields.

swolchok replied:
> Make the first word the size unconditionally; when size is greater than 5, stash the pointer in the inline fields. 

This gives us two places that store the size for the out-of-line case, adding implementation complexity. However, see the second benefit below!

> The benefit of the alternate encoding is that you can now retrieve the dim() of a tensor without any branches. 

I don't think that this benefit is worthwhile. A correctly predicted branch takes 1 cycle, and we expect much more than 99% of tensors to have rank 5 or below. We should expect a correct prediction on this the vast majority of the time.

That said, I can think of some other benefits you don't mention:
- We get to drop the bit shift to retrieve the size, which is nice to have and simplifies other code paths.
- Optimizing the out-of-line representation is a lot easier -- instead of a pointer to a vector, we can just use a pointer to an array of the right size directly, without worrying about where to put the size.

Overall, I think this is a good idea. After I'm sure that test coverage for the current version is good, I'll switch over to this suggestion and get rid of the `vector` as well. Thanks!

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -17,26 +30,118 @@ class C10_API SizesAndStrides {
+      } else {
+        // Both out of line. Move their storage into ours.
+        *outOfLineStorage() = std::move(*rhs.outOfLineStorage());
+        delete rhs.outOfLineStorage();
```
How come we can't just do `outOfLineStorage() = rhs.outOfLineStorage()`?

ezyang replied:
OK, true, need to delete `outOfLineStorage()` first. But besides that?

swolchok replied:
good point.

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -73,46 +194,120 @@ class C10_API SizesAndStrides {
+          const auto bytesToZero = (newSize - oldSize) * sizeof(inlineStorage_[0]);
+          memset(&inlineStorage_[oldSize], 0, bytesToZero);
+          memset(&inlineStorage_[MAX_INLINE_SIZE + oldSize], 0, bytesToZero);
+        }
```
In debug mode we should zero/sentinel out the slop, make it easier to figure out if something has gone wrong.

smessmer replied:
I would advocate for doing it in release mode too. Stuff being zero-initialized in debug mode but not in release mode is a long standing issue in some C++ compilers that makes debugging hard since stuff breaks when you run it in release mode and then suddenly it works when you rerun in debug mode.

swolchok replied:
I'd rather not add unnecessary memory references in release builds. AddressSanitizer will help with use of uninitialized memory already, and we can add test coverage to make sure things are working, so I've opted for not adding this debugging feature.

swolchok committed:
fix move-assign. delete unused method. vector-less representation coming in next update on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

smessmer reviewed (commented):
Not adding comments since you're planning a major rewrite. Let me know when that is finished so I can take a look.

robieta commented:
Sorry, https://github.com/pytorch/pytorch/pull/47507#issuecomment-724989045 was meant for this PR.

swolchok committed:
change to @ezyang's suggested representation. haven't gotten to do a cleanup yet on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
remove debug-only features on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
fix typo on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -19,18 +27,101 @@ class C10_API SizesAndStrides {
+  // Move from rhs. rhs.size() == 0 afterwards.
+  SizesAndStrides(SizesAndStrides&& rhs) : size_(rhs.size_) {
+    if (C10_LIKELY(isInline())) {
+      memcpy(inlineStorage_, rhs.inlineStorage_, sizeof(inlineStorage_));
```
Branch is saved, but when rhs is out of line, this is at the cost of copying extra words we don't need. Guessing it's better to save the branch.

ezyang reviewed (approved):
Did a reasonably close reading of resize and it looks good. Probably the one thing that's good to check on that ASAN/UBSAN won't catch you out on is if we're leaking memory; maybe running the test suite under LSAN should be good enough.

swolchok commented:
> maybe running the test suite under LSAN should be good enough.

ran with `ASAN_OPTIONS=print_suppressions=1:detect_leaks=1`, no failures

swolchok committed:
no reason to inline SizesAndStrides::resize on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
fix quoted import on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
add noexcept. check malloc/realloc return value on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
don't specify functions with TORCH_CHECK as noexcept on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
remove more noexcept specs on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
fix 2 typos in bottom diff on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
remove freeOutOfLineStorage so that operator= can be properly noexcept on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
rebase on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
add definition of static constexpr as required for C++14 on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Change representation of SizesAndStrides"


This moves SizesAndStrides to a specialized representation
that is 5 words smaller in the common case of tensor rank 5 or less.

Differential Revision: [D24772023](https://our.internmc.facebook.com/intern/diff/D24772023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24772023/)!

[ghstack-poisoned]
